### ISSUE TITLE
Safe Eyes 2.0.0

### ISSUE THREAD

**Issue by slgobinath** [slgobinath/SafeEyes#189]
Hi,
I am pleased to announce that Safe Eyes 2.0.0 is almost ready. However, I just need to confirm that merging it to `master` branch will not affect the build packages for Debian, Gentoo, and others (if there are any).

@hartwork  and @anarcat  could you please confirm this?

Also, I have changed the translation technique. Earlier I used JSON files but now I have switched to formal translation files. It requires all the existing translations to be updated to the new format. Some existing literals are removed and some new literals are added to the translation.

I kindly expect your contribution on translating Safe Eyes 2.0.0, updating the document and testing before releasing the final version. You can test Safe Eyes 2.0.0 by following these steps:

1. `git clone https://github.com/slgobinath/SafeEyes.git`

2. `cd SafeEyes`

3. `git checkout safeeyes-2.0.0`

4. Build the translation file using the following command:
```
msgfmt -o safeeyes/config/locale/en_US/LC_MESSAGES/safeeyes.mo safeeyes/config/locale/en_US/LC_MESSAGES/safeeyes.po
```

4. `python3 -m safeeyes`

Translations cannot be done via Weblate until I merge the changes to `master` branch. So please send your commits to the `safeeyes-2.0.0` branch. I left the previous JSON files for reference but they will be removed before releasing the final version.

**What is new in Safe Eyes 2.0.0:**
- Completely new architecture
- Features are implemented as plugins
- Plugins become more powerful
    - Plugins can have their own settings and the cool thing is: Safe Eyes automatically creates the respective Settings dialog.
    - Plugins can have their own dependencies
    - Plugins can control Safe Eyes or just display a text content in the break screen (Respective flow control methods are defined)
    - For more details, have a look at the existing plugins and the partially completed [README](https://github.com/slgobinath/SafeEyes/tree/safeeyes-2.0.0#writing-plug-in-for-safe-eyes) document
    - Users can create their own plugins. I am thinking about maintaining a directory/separate repo for custom plugins depending on the requirement.

- Users can manually start a break
- Persist the internal state
- All most all the configurations can be done via GUI

One more feature is pending to implement. Since the Tray Icon is implemented as `plugins`, it can be disabled by the user which prevents access to Settings Dialog. I let users disable it considering the Gnome 3 guidelines. However, I will implement an alternative way to launch the Settings dialog without using the Tray Icon. For the moment, stay away from disabling Tray Icon and in case if you have sone so, reset the settings by removing the `~/.config/safeeyes` folder.

If you have any suggestions, please do let me know.

![safe eyes](https://user-images.githubusercontent.com/4003168/31059676-44ae8f46-a6d4-11e7-8420-d59da1817436.jpg)

**Comment by anarcat** [comment]
>  I am pleased to announce that Safe Eyes 2.0.0 is almost ready. However, I just need to confirm that merging it to master branch will not affect the build packages for Debian, Gentoo, and others (if there are any).

shouldn't be a problem: safeeyes is in the "NEW" queue as 1.2.2 for now. a new version *could* be uploaded there by the maintainer, but it's mostly orthogonal to development here at this stage.

congrats on the release.

**Comment by hartwork** [comment]
Hi!

I only had time for a quick spin so far: It seems like the build system is not fully installing the plugins? During start-up it reads
```console
# safeeyes 
2017-10-02 20:03:25,619 [INFO]:[MainThread] Starting Safe Eyes
2017-10-02 20:03:25,648 [INFO]:[MainThread] Reading the configuration file
2017-10-02 20:03:25,649 [INFO]:[MainThread] Update the old config version 5.0.1 with new config version 6.0.0
2017-10-02 20:03:25,649 [INFO]:[MainThread] Copy the config files to ~/.config/safeeyes
2017-10-02 20:03:25,689 [INFO]:[MainThread] Initialize the break screen
2017-10-02 20:03:25,689 [INFO]:[MainThread] Load all the plugins
2017-10-02 20:03:25,689 [INFO]:[MainThread] Initialize the core
2017-10-02 20:03:25,689 [ERROR]:[MainThread] plugin.py not found for the plugin: notification
2017-10-02 20:03:25,690 [ERROR]:[MainThread] plugin.py not found for the plugin: trayicon
2017-10-02 20:03:25,690 [ERROR]:[MainThread] plugin.py not found for the plugin: screensaver
2017-10-02 20:03:25,690 [INFO]:[MainThread] Start Safe Eyes core
2017-10-02 20:03:25,690 [INFO]:[Thread-1] Waiting for 15 minutes until next break
```
and the `plugin.py` files seem missing, indeed. Expect this issue to affect packagin for other distros like Debian, as well.

**Comment by anarcat** [comment]
i noticed that my settings were lost after running 2.0 - is that normal?

furthermore, i was hoping to see improvements in the preferences dialog in the 2.0 release - but I have opened a separate issue about this in #190.

**Comment by slgobinath** [comment]
@hartwork ,
I haven't updated the `setup.py` to include the changes. I left packaging as the final part. I will update that soon.

@anarcat,
The configuration of Safe Eyes 2.0.0 highly differs from the config of Safe Eyes 1.2.x. Therefore I let it replace the entire config file.
One more thing to notice: if you return to Safe Eyes 1.2.x, please remvoe the config directory using the following command.

```
rm -r ~/.config/safeeyes
```
Please accept my apologies for the inconvenience.

Thanks

**Comment by anarcat** [comment]
okay well *that* should be an issue with debian :p we like our backwards-compatibility and warn users about stuff like that (in NEWS.Debian, for example). thanks for letting me know, i'll forward this to @FedericoCeratto who's working on the package.

**Comment by hartwork** [comment]
Maybe import from `~/.config/safeeyes` if found, leave unmodified, and use `~/.config/safeeyes2` for 2.x.x from that point on, if it's not too much work. It could smooth the upgrade path.

**Comment by slgobinath** [comment]
@hartwork & @anarcat ,
I will make the config directory name safeeyes2.

Thanks for the idea :-)

**Comment by anarcat** [comment]
another issue - after "pip install ." in the safeeyes 2.0 branch, it doesn't find the plugins properly:

```
[1006]anarcat@curie:~1$ safeeyes 
2017-10-02 20:54:07,894 [INFO]:[MainThread] Starting Safe Eyes
2017-10-02 20:54:07,908 [INFO]:[MainThread] Reading the configuration file
2017-10-02 20:54:07,908 [INFO]:[MainThread] Reading the session file
2017-10-02 20:54:07,961 [INFO]:[MainThread] Initialize the break screen
2017-10-02 20:54:07,962 [INFO]:[MainThread] Load all the plugins
2017-10-02 20:54:07,962 [INFO]:[MainThread] Initialize the core
2017-10-02 20:54:07,962 [ERROR]:[MainThread] plugin.py not found for the plugin: trayicon
2017-10-02 20:54:07,962 [INFO]:[MainThread] Start Safe Eyes core
2017-10-02 20:54:07,963 [INFO]:[Thread-1] Waiting for 15 minutes until next break
```

whereas when running from source, it is fine:

```
[1011]anarcat@curie:SafeEyes$ python3 -m safeeyes 
2017-10-02 20:54:26,601 [INFO]:[MainThread] Starting Safe Eyes
2017-10-02 20:54:26,614 [INFO]:[MainThread] Reading the configuration file
2017-10-02 20:54:26,618 [INFO]:[MainThread] Reading the session file
2017-10-02 20:54:26,677 [INFO]:[MainThread] Initialize the break screen
2017-10-02 20:54:26,678 [INFO]:[MainThread] Load all the plugins
2017-10-02 20:54:26,678 [INFO]:[MainThread] Initialize the core
2017-10-02 20:54:26,680 [INFO]:[MainThread] Successfully loaded <module 'trayicon.plugin' from '/home/anarcat/dist/SafeEyes/safeeyes/plugins/trayicon/plugin.py'>
2017-10-02 20:54:26,682 [INFO]:[MainThread] Successfully loaded <module 'healthstats.plugin' from '/home/anarcat/dist/SafeEyes/safeeyes/plugins/healthstats/plugin.py'>
2017-10-02 20:54:26,684 [INFO]:[MainThread] Successfully loaded <module 'screensaver.plugin' from '/home/anarcat/dist/SafeEyes/safeeyes/plugins/screensaver/plugin.py'>
2017-10-02 20:54:26,685 [INFO]:[MainThread] Successfully loaded <module 'donotdisturb.plugin' from '/home/anarcat/dist/SafeEyes/safeeyes/plugins/donotdisturb/plugin.py'>
2017-10-02 20:54:26,687 [INFO]:[MainThread] Successfully loaded <module 'audiblealert.plugin' from '/home/anarcat/dist/SafeEyes/safeeyes/plugins/audiblealert/plugin.py'>
2017-10-02 20:54:26,687 [DEBUG]:[MainThread] Initialize Tray Icon plugin
2017-10-02 20:54:26,693 [DEBUG]:[MainThread] Initialize Health Stats plugin
2017-10-02 20:54:26,693 [DEBUG]:[MainThread] Initialize Screensaver plugin
2017-10-02 20:54:26,693 [DEBUG]:[MainThread] Initialize Skip Fullscreen plugin
2017-10-02 20:54:26,693 [DEBUG]:[MainThread] Initialize Audible Alert plugin
2017-10-02 20:54:26,693 [INFO]:[MainThread] Start Safe Eyes core
2017-10-02 20:54:26,693 [INFO]:[Thread-1] Update next break information
2017-10-02 20:54:26,716 [INFO]:[Thread-1] Waiting for 15 minutes until next break

```

**Comment by svepe** [comment]
I was wondering what does the _Persist the internal state_ setting do?

**Comment by FedericoCeratto** [comment]
SafeEyes 1.2.2 is still in the upload queue and hasn't entered Debian just yet. If 2.0.0 is released soon I might be able to upload it instead of 1.2.2 and not have to handle the configuration transition.

**Comment by slgobinath** [comment]
@svepe it persists the next break information and starts from where it stops even after system restarts. If plugins support, it will save their states as well.

> SafeEyes 1.2.2 is still in the upload queue and hasn't entered Debian just yet. If 2.0.0 is released soon I might be able to upload it instead of 1.2.2 and not have to handle the configuration transition.

If so, I will release it within 24 hours. Thanks for notifying me.

**Comment by slgobinath** [comment]
@FedericoCeratto @anarcat and @hartwork,
I have a question. Is it recommended to translate command line help in Linux? According to my testings, most of the command line help outputs are not translated to my language. Further, if the terminal does not render the Unicode text properly, translated help looks messy.

**Comment by svepe** [comment]
@slgobinath @FedericoCeratto 
I have no idea if debian package dependencies are manually managed, but here is what I had to sort out when testing 2.0.0. Since the way audio is played has changed, the .deb package should no longer depend on `python3-pyaudio` but replace it with `alsa-utils`.

**Comment by FedericoCeratto** [comment]
@slgobinath distributions have been shipping localized CLI and graphical applications and manpages even before unicode using the locale toolset. There are libraries to detect the LANG, LC* env vars and replace strings accordingly.
(However for a graphical app like SafeEyes there's probably little need for a CLI)

**Comment by slgobinath** [comment]
> I have no idea if debian package dependencies are manually managed, but here is what I had to sort out when testing 2.0.0. Since the way audio is played has changed, the .deb package should no longer depend on python3-pyaudio but replace it with alsa-utils.
I have updated the dependency.

> distributions have been shipping localized CLI and graphical applications and manpages even before unicode using the locale toolset. There are libraries to detect the LANG, LC* env vars and replace strings accordingly.
(However, for a graphical app like SafeEyes there's probably little need for a CLI)

So I leave the command-line arguments as they are at the moment. Depending on requirements we'll consider them later.

**Comment by slgobinath** [comment]
Hi all,
Please note that I have released Safe Eyes 2.0.0.

**Comment by FedericoCeratto** [comment]
Thanks for the update. The new package has been prepared. #196 was detected during the build.

**Comment by radek-sprta** [comment]
In my experience, the well maintained software usually has command-line help localized. have also never had problems with unicode on command line with a properly configured locale.

-------- Original Message --------

**Comment by radek-sprta** [comment]
Congratultions on the release!

-------- Original Message --------

**Comment by mcarans** [comment]
A massive thumbs up for this application. :-) I have RSI and this really helps me to take breaks. Maybe you could consider changing the name as your utility is a general ergonomics utility not just for safeguarding eyes. Well done!

**Comment by hartwork** [comment]
> Please note that I have released Safe Eyes 2.0.0.

Bumped in Gentoo now: https://github.com/gentoo/gentoo/commit/38f0e09d2310046246383fe6c2f017e202ec024d

**Comment by slgobinath** [comment]
@hartwork thank you for the update.
